### ISSUE TITLE
Fix request: false positive on AvoidMutableStaticFields

### ISSUE THREAD

**Issue by stokpop** [jborgers/PMD-jPinpoint-rules#563]
Should we assume that classes (e.g. Avatar) are immutable on the inside?

The rule should only trigger on direct arrays, not on arrays given to methods or contructors. Agree?

Should we add/modify rule to check for non-copied arrays in constructors?

```java
class Foo {

    private static final Avatar defaultAvatar = new Avatar("application/png", new byte[]{-119, 80, 78}); // <-- false positive?

    private static final Avatar defaultAvatar2;

    static {
        final byte[] bytes = {-119, 80, 78};
        defaultAvatar2 = new Avatar("application/png", bytes); // no issue found here
        bytes[2] = 0;
    }


    private static class Avatar {
        private String contentType;
        private byte[] data;
        public Avatar(String contentType, byte[] bytes) {
            this.contentType = contentType;
            this.data = bytes; // should be a warning here instead to make defensive copy?
        }
        String getContentType() {
            return contentType;
        }
        byte[] getData() {
            return data;
        }
    }

    public static void main(String[] args) {
        // byte array is still modifiable on the outside
        defaultAvatar.data[0] = 0;
        System.out.println("byte one  :" + defaultAvatar.data[0]);
        // byte array modified on the outside after creation
        System.out.println("byte three:" + defaultAvatar2.data[2]);
    }
}
```

**Comment by stokpop** [comment]
Other false positive example, caused by `new CLOptionDescriptor[]` array

```java
    private static final CLOptionDescriptor D_REPORT_GENERATING_OPT = // <-- false positive
            new CLOptionDescriptor("reportonly",
                    CLOptionDescriptor.ARGUMENT_REQUIRED, REPORT_GENERATING_OPT,
                    "generate report dashboard only, from a test results file",
                    new CLOptionDescriptor[]{ D_NONGUI_OPT, D_REMOTE_OPT, D_REMOTE_OPT_PARAM, D_LOGFILE_OPT }); // disallowed
```

note that a defensive copy is made in this case:

```java
    public CLOptionDescriptor(final String name, final int flags, final int id, final String description,
                              final CLOptionDescriptor[] incompatible) {

        checkFlags(flags);

        this.id = id;
        this.name = name;
        this.flags = flags;
        this.description = description;

        this.incompatible = new int[incompatible.length];
        for (int i = 0; i < incompatible.length; i++) {
            this.incompatible[i] = incompatible[i].getId();
        }
    }
```
